add index priority pep draft (#5)
wheelnext/peps: peps/pep-9999.rst

PEP: 9999

> Title: Standardizing selection of Python package files across multiple sources

Reviewer: I think we can come up with a snazzier title that both makes the PEP easier to find, and captures the audiences attention.  I'd like to at least add the word "priority" and am not sure we need "Standarizing" as that's what the PEP process is inherently all about.

Author: See what you think. I'm not sure we'll make it to "snazzy," but I hope this is closer

Author: Michael Sarahan, msarahan@gmail.com
Sponsor: Barry Warsaw, barry@python.org
PEP-Delegate: <PEP delegate’s real name>
Discussions-To: <REQUIRED: URL of current canonical discussion thread>
Status: Draft

> Type: Standards Track

Reviewer: It seems like this could either be a [Standards Track or Information PEP](https://peps.python.org/pep-0001/#pep-types).  While it's proposing a standard, that standard isn't directly for Python, and tool implementers can follow the advice or not.  I'd lean more towards Informational, but if you feel strongly about ST, then that's fine with me.

Author: I went with Information. Thanks.

Topic: Packaging
Requires: 777
Created: 05-Nov-2024
Post-History: <REQUIRED: dates, in dd-mmm-yyyy format, and corresponding links to PEP discussion threads>

Abstract
========

> Python installers are diverging with respect to how they handle multiple

Reviewer 2: I feel like this hook dives really deep into the meat of the topic without setting proper context for why someone should care. I might recommend giving some context about package resolution and why it is important. That should better motivate why having the same (or at least similar) behavior would be useful.

Author: I made an attempt here. See what you think.

indexes. Pip merges all indexes before matching distributions, while uv
matches distributions on one index before moving on to the next. Each
approach has advantages and disadvantages. This PEP aims to standardize
each of these behaviors, which are referred to as “version priority” and

> “index priority” respectively, so that community discussions and

Reviewer 2: Nit: I think the reasoning is much more than just to standardize community discussions. I would be explicit and say the goals are both to standardize terms used when discussing index priority, and standardize behavior of tools to make the behavior consistent between them.

Reviewer 2: You do say this below in the "Goals" section, but I think it would be good to mention it here too.

will also facilitate building tools that rely on these behaviors.

Motivation
==========

Goals
-----

- Establish standard for evaluation order of various levels of priority
  between indexes and packages
- Standardize verbiage of different approaches, so that cross-tool
  discussions are talking “apples to apples”
- Suggest standard configuration options, to minimize user cognitive
  load between tools
- Provide guidelines for how ecosystem tools should implement index
  priority if they would like, and suggest reasons why they might want
  to do so

Non-goals
---------

- Do not relax the simplifying assumption that for a given distribution
  filename, all indexes must have the same file. This has been expressed
  as “all indexes are equal priority”, but can be more accurately
  described as “`pip considers all distributions with the same name and
  version to be interchangeable regardless of which index they came
  from <https://github.com/astral-sh/uv/issues/171#issuecomment-1952079681>`__.”
  This is a key assumption for how pip behaves today, and changing this
  assumption is deemed too disruptive for this PEP.

> - Do not promote index priority as a solution to dependency confusion.

Reviewer: This PEP might not be a full solution to the dependency confusion problem, but it can contribute to a safer ecosystem.  I'm also not personally optimistic that PEP 708 is realistic, given its provisional nature and reliance on implementation in other index platforms.

I suppose my recommendation would be to perhaps move this to the "Goals" but couch it in terms like improvements but not solutions.

Author: Please check rewording here

  It is related, but `PEP 708 <https://peps.python.org/pep-0708/>`__ is
  a better, more specific solution that does not require user
  configuration
- Do not dictate that every tool must implement every strategy

Rationale
=========

> This PEP proposes standards of installer behavior when using multiple

Reviewer: As @ethanhs-nv points out, an important rationale is to align on terminology, and I think "version priority" and "index priority" are good terms for the algorithms (which is one reason to get the word "priority" in the title).

approaches. Uv and PDM have each added support for some notion of index
priority.

Index priority is `the default behavior in
uv <https://docs.astral.sh/uv/pip/compatibility/#packages-that-exist-on-multiple-indexes>`__,
and supported, but `not default behavior in
PDM <https://pdm-project.org/latest/usage/config/#respect-the-order-of-the-sources>`__.
In uv, `users have reported some
issues <https://github.com/astral-sh/uv/issues/2775>`__ with the
behavior and asked for keeping pip’s behavior. Uv provides this choice
as a ``--index-strategy`` flag, with the ``unsafe-best-match`` value
matching pip’s current behavior, and the ``first-match`` value matching
this PEP’s proposed behavior. Defining index priority is expected to
help with `situations where a local package has remote
dependencies <https://github.com/pypa/pip/issues/11624>`__, and the user
wishes to prioritize local packages over remote dependencies, while
still falling back to remote dependencies where needed.

The topic of prioritizing indexes is being proposed as a PEP rather than
as a change specifically to pip, because predictable index behavior
across the many tools in our ecosystem is important for consistent,

> predictable behavior between tools.

Reviewer 2: ```suggestion
predictable package resolution behavior between tools.
```

Specification
=============

“Version priority”
------------------

This is pip’s current behavior. All distributions from all configured

> repositories are collated, and the “best” among them is selected. The

Reviewer: ```suggestion
[configured] and enabled indexes are collated, and the “best” among them is selected. The
```

- `Absence of
  yanking <https://github.com/pypa/pip/blob/111eed14b6e9fba7c78a5ec2b7594812d17b5d2b/src/pip/_internal/index/package_finder.py#L542>`__
- `Wheels over
  sdists <https://github.com/pypa/pip/blob/111eed14b6e9fba7c78a5ec2b7594812d17b5d2b/src/pip/_internal/index/package_finder.py#L504>`__
- `Maximize
  version <https://github.com/pypa/pip/blob/111eed14b6e9fba7c78a5ec2b7594812d17b5d2b/src/pip/_internal/index/package_finder.py#L544>`__
  (including `Local version
  tags <https://peps.python.org/pep-0440/#local-version-identifiers>`__,
  `restricted to locally patched private
  versions <https://discuss.python.org/t/lets-permit-local-version-label-in-version-specifiers/22781/2>`__)
- `Maximize specificity of platform
  tag <https://github.com/pypa/pip/blob/111eed14b6e9fba7c78a5ec2b7594812d17b5d2b/src/pip/_internal/index/package_finder.py#L520>`__
- `Build
  tags <https://github.com/pypa/pip/blob/111eed14b6e9fba7c78a5ec2b7594812d17b5d2b/src/pip/_internal/index/package_finder.py#L535>`__
  with the first value being an integer interpreted as ascending values
  being “better”, and the second being a string with lexical sorting

> This behavior is characterized by the installer always getting the

Reviewer: Perhaps move this to above the bulleted list?  I think this is a good summary of what "version priority" means, so lead with that?

latest version of a package, regardless of the repository that it comes

> from. It is especially useful when all configured repositories are

Reviewer: And what if those indexes are _not_ equally trusted?  Should version priority be avoided entirely?  Perhaps say something about that case?

Author: Added stuff here, but it felt clumsy. See what you think.

hardware, and this version priority behavior can lead to undesirable,
unexpected outcomes.

Uv refers to `its implementation of the version priority
behavior <https://docs.astral.sh/uv/pip/compatibility/#packages-that-exist-on-multiple-indexes>`__
as “``unsafe-best-match``.” Naming is really hard here. On one hand, it
isn’t accurate to call pip’s default behavior intrinsically “unsafe.”
The addition of possibly malicious indexes/repositories is what
introduces concern with this behavior. PEP 708 added a way to restrict
installers from drawing packages from unexpected, potentially insecure
indexes. On the other hand, the term “best-match” is technically
correct, but also misleading. The “best match” varies by user and by
application. “Best” is technically correct in the sense that it is a
global optimum according to the match criteria specified above, but that
is not necessarily what is “best” in a user’s eyes. “Version priority”
is a proposed term that avoids the concerns with the uv terminology,
while approximating the behavior in the most user-identifiable way that
packages are compared.

“Index priority”
----------------

> Relative to “version priority” behavior, this strategy does not collate

Reviewer: What about adding a leading summary which describes index priority without being relative to version priority?

Author: Added. Please see what you think.

distribution. The index (or “source” in PDM terms) priority is
explicitly defined by a configuration hierarchy. Each package’s finder
starts at the beginning of the list of repositories, so each package
starts over with the repository list. In other words, if one package has
no valid candidates on the first index, but finds a hit on the second
index, subsequent packages still start their search on the first index,
rather than starting on the second.

One desirable behavior that the index priority strategy implies is that
there are no “surprise” updates, where a version bump on a
lower-priority repo wins out over a curated, approved higher-priority
repo. This is related to the security improvement of PEP 708, where
packages can restrict the external repos that distributions can come
from, but it is more configurable by end users. The package installs are
only expected to change when either the higher-priority repo or the
index priority configuration change. This stability and predictability
makes it viable to configure indexes as a more persistent property of an
environment, rather than a one-off argument for one install command.

> One important implementation detail of index priority is that caching

Reviewer: Don't lockfiles have the same implications?

Author: Definitely. Added.

should now include the index from which distributions were downloaded.
Without this aspect, it is possible that after changing the list of
configured indexes, the cache could provide a similarly-named
distribution from a lower-priority index. If every index follows the
recommended behavior of providing identical files across indexes for a
given filename, this is not an issue. However, that recommendation is
not readily enforceable, and augmenting the cache key with origin index
would be a wise defensive change.

Ways that a request falls through to a lower priority index
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Package name is not present at all in higher priority index
- All distributions from higher priority index filtered out due to

>   version specifier, python version, platform tag, yanking or otherwise

Reviewer: ```suggestion
  version specifier, compatible Python version, platform tag, yanking or otherwise
```

::

   --extra-index-url <req_file_index_url_1>
   --find-links <some_local_path>
   --extra-index-url <req_file_index_url_2>
   my-package

The CLI command of:

::

   pip install --extra-index-url <something> -r requirements.txt --extra-index-url <another>

Would expand to an index priority list of:

1. <some_local_path>
2. <something>
3. <req_file_index_url_1>
4. <req_file_index_url_2>
5. <another>
6. <fallback to value of index-url>

Because the requirements.txt file is parsed from top to bottom, it
evaluates ``--extra-index-url`` and ``--find-links`` arguments in the
order that they appear in the requirements.txt file, as opposed to
reordering them such that ``--find-links`` arguments always come before
any ``--extra-index-url`` arguments. Requirements files may also include
other requirements files. In the general case, depth-first
traversal of requirements files defines the collection and ordering
of index-urls and find-links options.

Proposed deny list format
~~~~~~~~~~~~~~~~~~~~~~~~~

As described above, `it may be desirable to intentionally omit one or
more packages from consideration on a given
index <https://github.com/astral-sh/uv/issues/4753>`__. For example, if

> a repository includes builds of a package that introduces issues, but

Reviewer: The PEP uses "repository" and "index" somewhat interchangeably and inconsistently, but I think it's better to converge on the term "index" for what PyPI-like things are.

Author: I replaced all instances of repo with index. I agree that loosey-goosey flipping is not helpful here.

  verbosely, because it is counter to pip’s well-established behavior.
- Improving the predictability of which repository will be selected may
  tempt people into using index priority as a way of having similarly
  named files that have different contents. It would be helpful if tools
  errored on mismatching hashes, or otherwise gave the user feedback
  that they are breaking a key assumption. This PEP does not mandate
  alternative fixes, such as augmenting the cache key with the index,
  because these changes have unknown unintended consequences. However,
  it is advisable to augment the cache key with the index, as doing so
  would remedy known “gotchas.”

Security Implications
=====================

Index priority creates a mechanism for users to explicitly specify a
trust hierarchy among their indexes. As such, it limits the potential
for dependency confusion attacks. Index priority was `rejected by PEP
708 <https://peps.python.org/pep-0708/#provide-a-mechanism-to-order-the-repositories>`__
as a solution for dependency confusion attacks. This PEP requests that
the rejection be reconsidered, with index priority serving a different

> purpose. This PEP is primarily motivated by the desire to support

Reviewer: See, this PEP _does_ kind of promote one of its anti-goals 😄 - which I think is a good thing!

implementation variants, which is the subject of `another discussion
that hopefully leads to a
PEP <https://discuss.python.org/t/selecting-variant-wheels-according-to-a-semi-static-specification/53446>`__.
It is not mutually exclusive with PEP 708, nor does it suggest reverting

> or withdrawing PEP 708. It is an answer to “\ `how we could allow users

Reviewer: ```suggestion
or withdrawing PEP 708. It is an answer to `how we could allow users
```

supported behaviors need documentation, and we, the authors of this
PEP, would develop these as part of the review period of this PEP.
These docs would likely consist of additions across several
repositories, cross-linking the concepts between installers. At a
minimum, we expect to add to the
`PyPUG <https://packaging.python.org/en/latest/>`__ and to `pip’s
documentation <https://pip.pypa.io/en/stable/cli/pip_install/>`__.

It will be important to advertise the active behavior, especially in
error messaging, and that will provide ways to provide resources to
users about these behaviors.

Uv users are already experiencing index priority. Uv `documents this
behavior <https://docs.astral.sh/uv/pip/compatibility/#packages-that-exist-on-multiple-indexes>`__
well, but it is always possible to `improve the
discoverability <https://github.com/astral-sh/uv/issues/4389>`__ of that
documentation from the command line, `where users will actually
encounter the unexpected
behavior <https://github.com/astral-sh/uv/issues/5146>`__.

Reference Implementation
========================

The uv project demonstrates index priority with its default behavior. Uv

> is implemented in Rust, though, so if a python reference implementation

Reviewer: ```suggestion
is implemented in Rust, though, so if a reference implementation to a Python-based tool
```

is necessary, we, the authors of this PEP, will provide one. For pip in
particular, we see the implementation plan as something like:

- For users who don’t use ``--extra-index-url`` or ``--find-links``,
  there will be no change, and no migration is necessary.
- Pip users would be able opt in to the index priority behavior with a

>   new config setting in the CLI and in pip.conf. This proposal does not

Reviewer: ```suggestion
  new config setting in the CLI and in ``pip.conf``. This proposal does not
```

  recommend any strategy as the default for any installer. It only
  recommends documenting the strategies that a tool provides.
- Enable extra info-level output for any pip operation where more than
  one index is used. In this output, state the current strategy setting,
  and a terse summary of implied behavior, as well as a link to docs
  that describe the different options
- Add debugging output that verbosely identifies the index being used at
  each step, including where the file is in the configuration hierarchy,
  and where it is being included (via config file, env var, or CLI
  flag).
- Plumb tracking of which repository gets used for which
  package/distribution through the entire pip install process. Store
  this information so that it is available to tools like ``pip freeze``

> - Supplement PEP 751 (lockfiles) with capture of repository where a

Reviewer: ```suggestion
- Supplement :pep:`751` (lockfiles) with capture of repository where a
```

  package/distribution came from

Rejected Ideas
==============

> - Tell users to set up a proxy/mirror, such as devpi or artifactory that

Reviewer: Links to "devpi" and "artifactory"?
